src/scripts/interactions.js

```diff
@@ -0,0 +1,216 @@
+    if (ray) {
+      for (const intersection of raycast(ray)) {
+        func(intersection, inputSource, new Matrix4().fromArray(ray.matrix));
+        break;
+      }
+    }
+  };
+}
+// Each input source can only be dragging one thing at a time, so:
+// dragAndDrop is a map from inputSource -> {offset}
+const dragAndDrop = new Map();
+const handleSelectStart = handlerCommon((intersection, inputSource, pointerMatrix) => {
+  const interactions = intersection.object[Interactions];
+  if (interactions) {
+    if (interactions.drag_start || interactions.drag_end || interactions.drag) {
+      // TODO: Apply rotation
+      let data;
+      if (interactions.drag_start) {
+        data = interactions.drag_start(intersection, pointerMatrix);
+      } else {
+        const pointerInverse = new Matrix4().getInverse(pointerMatrix, true);
+        const target = new Matrix4().copy(intersection.object.matrixWorld);
+        const transformMatrix = new Matrix4().multiplyMatrices(pointerInverse, target);
+        data = {
+          object: intersection.object,
+          transformMatrix,
+          matrixAutoUpdate: intersection.object.matrixAutoUpdate
+        };
+      }
+      intersection.object.matrixAutoUpdate = false;
+      dragAndDrop.set(inputSource, data);
+    } else if (interactions.select_start) {
+      interactions.select_start(intersection);
+    }
+  }
+});
+const handleSelectEnd = handlerCommon((intersection, inputSource) => {
+  // Handle the end of dragging
+  const data = dragAndDrop.get(inputSource);
+  if (data) {
+    const dragend = data.object[Interactions].drag_end;
+    data.object.matrixAutoUpdate = data.matrixAutoUpdate;
+    dragAndDrop.delete(inputSource);
+    if (dragend) {
+      dragend();
+    }
+  }
+
+  const interactions = intersection.object[Interactions];
+  if (interactions) {
+    // Handle the end of selection
+    if (interactions.select_end) {
+      interactions.select_end(intersection);
+    }
+  }
+});
+const handleSelect = handlerCommon((intersection) => {
+  const interactions = intersection.object[Interactions];
+  if (interactions) {
+    if (interactions.select) {
+      interactions.select(intersection);
+    }
+  }
+});
+
+export function setupInteractions() {
+  inputSources = XR.session.getInputSources();
+  console.log(inputSources);
+  XR.session.addEventListener('inputsourceschange', handleInputSourcesChange);
+
+  XR.session.addEventListener('select', handleSelect);
+  XR.session.addEventListener('selectstart', handleSelectStart);
+  XR.session.addEventListener('selectend', handleSelectEnd);
+}
+
+const raycaster = new Raycaster();
+function raycast(xrRay) {
+  const { scene } = getCurrentScene();
+
+  const trMatrix = new Matrix4().fromArray(xrRay.matrix);
+
+  // Transformed ray matrix from the current scene matrix world
+  const rMatrix = new Matrix4().multiplyMatrices(scene.matrixWorld, trMatrix);
+
+  raycaster.set(
+    new Vector3().setFromMatrixPosition(rMatrix),
```

Comment:
Is the user position being included here? Its hard to tell without the controller ray being rendered in yet.

Reply:
I think with the originOffset we won't have to worry about translating things around anymore.  The only thing I've had to adjust was inverting the teleportation coordinates before editing the offsetMatrix.